packages/core/src/routes/session/forgot-password.ts

```diff
@@ -0,0 +1,66 @@
+import { emailRegEx, phoneRegEx } from '@logto/core-kit';
+import { PasscodeType } from '@logto/schemas';
+import { Provider } from 'oidc-provider';
+import { z } from 'zod';
+
+import RequestError from '@/errors/RequestError';
+import { createPasscode, sendPasscode } from '@/lib/passcode';
+import koaGuard from '@/middleware/koa-guard';
+import { hasUserWithEmail, hasUserWithPhone } from '@/queries/user';
+import assertThat from '@/utils/assert-that';
+
+import { AnonymousRouter } from '../types';
+import { getRoutePrefix } from './utils';
+
+export const forgotPasswordRoute = getRoutePrefix('forgot-password');
+
+export default function forgotPasswordRoutes<T extends AnonymousRouter>(
+  router: T,
+  provider: Provider
+) {
+  router.post(
+    `${forgotPasswordRoute}/sms/send-passcode`,
+    koaGuard({ body: z.object({ phone: z.string().regex(phoneRegEx) }) }),
+    async (ctx, next) => {
+      const { jti } = await provider.interactionDetails(ctx.req, ctx.res);
+      const { phone } = ctx.guard.body;
+      const type = 'ForgotPasswordSmsSendPasscode';
+      ctx.log(type, { phone });
+
+      assertThat(
+        await hasUserWithPhone(phone),
+        new RequestError({ code: 'user.phone_not_exists', status: 422 })
+      );
```

Comment:
per our discussion, no need to guard the user existence here.

Reply:
updated